pkg/icingadb/ha.go

```diff
@@ -370,23 +376,48 @@ func (h *HA) realize(
 
 			if takeover != "" {
 				stmt := h.db.Rebind("UPDATE icingadb_instance SET responsible = ? WHERE environment_id = ? AND id <> ?")
-				_, err := tx.ExecContext(ctx, stmt, "n", envId, h.instanceId)
+				if _, err := tx.ExecContext(ctx, stmt, "n", envId, h.instanceId); err != nil {
+					return database.CantPerformQuery(err, stmt)
+				}
 
-				if err != nil {
+				// Insert the environment after each heartbeat takeover if it does not already exist in the database
+				// as the environment may have changed, although this is likely to happen very rarely.
+				stmt, _ = h.db.BuildInsertIgnoreStmt(h.environment)
+				if _, err := h.db.NamedExecContext(ctx, stmt, h.environment); err != nil {
 					return database.CantPerformQuery(err, stmt)
 				}
 			}
 
-			if err := tx.Commit(); err != nil {
-				return errors.Wrap(err, "can't commit transaction")
+			// In general, cancellation does not work for commit and rollback. Some database drivers may support a
+			// context-based abort, but only if the DBMS allows it. This was also discussed in the initial PR adding
+			// context support to sql: https://github.com/golang/go/issues/15123#issuecomment-245882486
+			//
+			// The following is implementation knowledge, not covered by the API specification. Go's sql.Tx.Commit() -
+			// which is not being overridden by sqlx.Tx - performs a preflight check on the context before handing over
+			// to the driver's Commit method. The drivers may behave differently. For example, the used
+			// github.com/go-sql-driver/mysql package calls its internal exec method with a "COMMIT" query, writing and
+			// reading packets without honoring the context.
+			//
+			// In a nutshell, one cannot expect a Commit() call to be covered by the transaction context. Thus, an
+			// additional check wraps the call, ensuring that this function will be left when the context is done.
```

Comment:
Can you please also add a comment that it is theoretically no problem to call realize again while the COMMIT is still running in the go routine.

Reply:
I wrote more prose. Please take another look.

Reply:
Looks good, ty! 👍